crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+{
+    if (ctx == NULL || privp == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->mode != OSSL_HPKE_MODE_AUTH
+        && ctx->mode != OSSL_HPKE_MODE_PSKAUTH) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->authpriv != NULL)
+        EVP_PKEY_free(ctx->authpriv);
+    ctx->authpriv = EVP_PKEY_dup(privp);
+    if (ctx->authpriv == NULL)
+        return 0;
+    return 1;
+}
+
+/**
+ * @brief set a public key for HPKE authenticated modes
+ * @param ctx is the pointer for the HPKE context
+ * @param pub is an buffer form of the public key
+ * @param publen is the length of the above
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_authpub(OSSL_HPKE_CTX *ctx,
+                               const unsigned char *pub, size_t publen)
+{
+    if (ctx == NULL)
+        return 0;
+    if (ctx->authpub != NULL)
+        OPENSSL_free(ctx->authpub);
+    ctx->authpub = OPENSSL_malloc(publen);
+    if (ctx->authpub == NULL)
+        return 0;
+    memcpy(ctx->authpub, pub, publen);
+    ctx->authpublen = publen;
+    return 1;
+}
+
+/**
+ * @brief ask for the state of the sequence of seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq returns the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the most recent used when sealing
+ * or opening (successfully)
+ */
+int OSSL_HPKE_CTX_get0_seq(OSSL_HPKE_CTX *ctx, uint64_t *seq)
+{
+    if (ctx == NULL || seq == NULL)
+        return 0;
+    *seq = ctx->seq;
+    return 1;
+}
+
+/**
+ * @brief set the sequence value for seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq set the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the most recent used when sealing
+ * or opening (successfully)
+ */
+int OSSL_HPKE_CTX_set1_seq(OSSL_HPKE_CTX *ctx, uint64_t seq)
+{
+    if (ctx == NULL)
+        return 0;
+    ctx->seq = seq;
+    return 1;
+}
+
+/**
+ * @brief sender seal function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param ct is the ciphertext output
+ * @param ctlen is the size the above
+ * @param pub is the recipient public key octets
+ * @param publen is the size the above
+ * @param infolen is the size the above
+ * @param info is the key schedule info parameter
+ * @param infolen is the size the above
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @param aad is the aad parameter
+ * @param aadlen is the size the above
+ * @param pt is the plaintext
+ * @param ptlen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * This can be called multiple times
+ */
+int OSSL_HPKE_sender_seal(OSSL_HPKE_CTX *ctx,
```

Comment:
Do we even need this interface at all?

Reply:
I left in all the previous APIs in this commit but will shortly drop 'em once I'm done with the stuff with tables etc.

Reply:
those are gone now